### PR TITLE
Alerting: Introduce the silencing interface

### PR DESCRIPTION
**Pull request by gotjosh**
The operations introduced are:

- Listing silences
- Retrieving an specific silence
- Deleting a silence
- Creating a silence

Signed-off-by: Josue Abreu <josue@grafana.com>

**Special notes for your reviewer**:

I have not not the integration point of API <> Silencing Interface but I'll address this is in a separate PR so that I can include some tests at that level.

### CODE REVIEW

File: pkg/services/ngalert/api/api_alertmanager_base.go [hunk 1]
```diff
@@ -18,14 +18,15 @@ import (
 )
 
 type AlertmanagerApiService interface {
-	RouteCreateSilence(*models.ReqContext, apimodels.SilenceBody) response.Response
```
Comment (gotjosh) [inline pkg/services/ngalert/api/api_alertmanager_base.go:21]: The big diff here is mostly due to a re-order, I've made it a bit easier to understand by it.

File: go.mod [hunk 2]
```diff
@@ -11,6 +11,8 @@ replace github.com/denisenkom/go-mssqldb => github.com/grafana/go-mssqldb v0.0.0
 // It's also present on grafana/loki's go.mod so we'll need till it gets updated.
 replace k8s.io/client-go => k8s.io/client-go v0.18.8
 
+replace github.com/prometheus/alertmanager => github.com/gotjosh/alertmanager v0.21.1-0.20210329182206-9ab7cef7510e
```
Comment (gotjosh) [inline go.mod:14]: Hopefully this gets merged before this PR is reviewed and I don't have to use my own fork.

File: pkg/services/ngalert/notifier/alertmanager.go [hunk 3]
```diff
@@ -33,6 +30,8 @@ import (
 
 const (
 	workingDir = "alerting"
+	// How long should we keep silences and notification entries on-disk after they've served their purpose.
+	retentionNotificationsAndSilences = 5 * 24 * time.Hour
```
Comment (gotjosh) [inline pkg/services/ngalert/notifier/alertmanager.go:34]: At this point, I prefer a set of sensible defaults as opposed to extra knobs.

**Review by papagian — APPROVED** [review]
AFAICT LGTM

**Pull request merged**